### ISSUE TITLE
Files with no extension are skipped

### ISSUE THREAD

**Issue by glubsy** [thumbsup/thumbsup#213]
## Problem

Files without an extension are being silently skipped during scan. For example:
```
$ ls ./my/path
file01.jpg
file01
file02.jpg
file03
```
Both `file01` and `file03` are missing from the gallery, and thumbsup doesn't seem to report that in the logs. 
However, they should be in the gallery since they are actually webm files in this case.

## Solution
We could detect the file type with packages such as [file-type](https://github.com/sindresorhus/file-type), which apparently does the same as a `$ file filename.ext` in Unix-like operating systems.

Note that `file01.jpg` and `file01` might end up having the same thumbnail file, file01.jpg which would be a problem. I know you have dealt with a [similar issue](https://github.com/thumbsup/thumbsup/issues/111) before. Perhaps thumbnails should keep the full filename with both name+extension -> `file01.jpg.jpg`?

## Extra
Optionally, for files which cannot be processed due to unsupported or unknown formats (ie. .swf, .zip, .rar, .txt, etc.), they should still be added to the gallery with some [default icon as the thumbnail](https://www.iconfinder.com/icons/25041/executable_gnome_fs_binary_icon). And at the very least should be reported in the log.

**Comment by rprieto** [comment]
Hi, thanks for raising this and providing a lot of detail. 

At the moment thumbsup searches for media files with glob patterns. It's technically possible to scan every single file instead. However it would have performance implications both for the indexing phase and the database size. To understand the use case, is there something preventing you from simply naming the file `file01.webp`?

Regarding non-photo files, what is your use case? Thumbsup is a gallery generator and does not intend to become a generic "file explorer". It would be interesting for sure, but I prefer to focus the complexity on gallery features such as metadata, geolocation, maps, etc.

Today, files that match the glob but cannot be processed (neither a photo or a video) are listed in the logs.

**Comment by glubsy** [comment]
Unsupported files are indeed reported if they pass the "allowed extensions" filter. A **SWF** file named **file.swf** will not get reported, unless it's named **file.mp4**, in which case it will then be parsed and reported as unsupported. 
It turns out that files lying about their actual format (with the wrong extension) are quite common.
Extension-less files are never reported since they get filtered out altogether (they do not pass [the filter](https://github.com/thumbsup/thumbsup/blob/master/src/components/index/glob.js#L24)).

For the sake of simplicity, let's just say that I cannot rename any file until after the gallery has been generated. My use case is to generate an overview of the files available in a read-only archive to share with colleagues. Any (renaming) operation on the file system should be postponed as late as possible.

Let's put aside the issue of non-supported files, which should have been a separate issue anyway (sorry about that). I reckon I probably need a proper [file-explorer](https://larsjung.de/h5ai/) but I couldn't find one that serves static files only.

A quick hack and extension-less files are now included:
```patch
diff --git src/components/index/pattern.js src/components/index/pattern.js
index aaa5456..5305914 100644
--- src/components/index/pattern.js
+++ src/components/index/pattern.js
@@ -10,9 +10,10 @@ class GlobPattern {

   match (filePath) {
     const opts = { nocase: true }
-    return micromatch.any(filePath, this.includeList, opts) &&
+    return (micromatch.any(filePath, this.includeList, opts) &&
            micromatch.any(filePath, this.excludeList, opts) === false &&
-           micromatch.isMatch(filePath, this.extensions, opts)
+           micromatch.isMatch(filePath, this.extensions, opts)) ||
+           filePath.indexOf('.') == -1
   }

   // this is used to cull folders early
```
However filenames with a period somewhere in them but no extension are still a problem. For example: "**random.filename**" which is actually a jpeg. 

Completely removing the extension check in the filter works just fine, files are still analyzed correctly afterwards as far as I can tell. 
```patch
diff --git src/components/index/pattern.js src/components/index/pattern.js
index aaa5456..e9753e2 100644
--- src/components/index/pattern.js
+++ src/components/index/pattern.js
@@ -11,8 +11,7 @@ class GlobPattern {
   match (filePath) {
     const opts = { nocase: true }
     return micromatch.any(filePath, this.includeList, opts) &&
-           micromatch.any(filePath, this.excludeList, opts) === false &&
-           micromatch.isMatch(filePath, this.extensions, opts)
+           micromatch.any(filePath, this.excludeList, opts) === false
   }
 
   // this is used to cull folders early
```
Perhaps this could warrant a new `--ignore-extensions` option to more aggressively include all files at the expense of a slightly slower scan?
I think it's safe to say that users will usually scan set of files which they know are mostly valid media files, so in practical terms I doubt this would make much difference performance-wise.

**Comment by glubsy** [comment]
Here is my current patch to include all supported files, regardless of extension.
```patch
diff --git bin/options.js bin/options.js
index 1413e19..e7747fd 100644
--- bin/options.js
+++ bin/options.js
@@ -30,22 +30,28 @@ const OPTIONS = {
   // ------------------------------------
   'include-photos': {
     group: 'Input options:',
-    description: 'Include photos in the gallery',
+    description: 'Include photos in the gallery (by common file extensions)',
     type: 'boolean',
     'default': true
   },
   'include-videos': {
     group: 'Input options:',
-    description: 'Include videos in the gallery',
+    description: 'Include videos in the gallery (by common file extensions)',
     type: 'boolean',
     'default': true
   },
   'include-raw-photos': {
     group: 'Input options:',
-    description: 'Include raw photos in the gallery',
+    description: 'Include raw photos in the gallery (by common file extensions)',
     type: 'boolean',
     'default': false
   },
+  'ignore-extensions': {
+    group: 'Input options:',
+    description: 'Include all supported files regardless of file extensions',
+    type: 'boolean',
+    'default': true
+  },
   'include': {
     group: 'Input options:',
     description: 'Glob pattern of files to include',
diff --git src/components/index/glob.js src/components/index/glob.js
index e77bfa8..557793c 100644
--- src/components/index/glob.js
+++ src/components/index/glob.js
@@ -34,6 +34,7 @@ exports.find = function (rootFolder, options, callback) {
 
 exports.supportedExtensions = function (options) {
   const extensions = []
+  if (options.ignoreExtensions == true) return extensions
   if (options.includePhotos !== false) Array.prototype.push.apply(extensions, PHOTO_EXT)
   if (options.includeVideos !== false) Array.prototype.push.apply(extensions, VIDEO_EXT)
   if (options.includeRawPhotos) Array.prototype.push.apply(extensions, RAW_PHOTO_EXT)
diff --git src/components/index/pattern.js src/components/index/pattern.js
index aaa5456..e11758c 100644
--- src/components/index/pattern.js
+++ src/components/index/pattern.js
@@ -12,7 +12,8 @@ class GlobPattern {
     const opts = { nocase: true }
     return micromatch.any(filePath, this.includeList, opts) &&
            micromatch.any(filePath, this.excludeList, opts) === false &&
-           micromatch.isMatch(filePath, this.extensions, opts)
+           (this.extensions === undefined ? true :
+            micromatch.isMatch(filePath, this.extensions, opts))
   }
 
   // this is used to cull folders early
@@ -28,9 +29,11 @@ class GlobPattern {
 function extPattern (extensions) {
   if (extensions.length === 1) {
     return '**/*.' + extensions[0]
-  } else {
+  } else if (extensions.length > 1) {
     return '**/*.{' + extensions.join(',') + '}'
   }
+  else
+    return undefined
 }
 
 module.exports = GlobPattern
```
However, this overrides the `--include-photos`, `--include-videos` and `--include-raw-photos` options, and these cannot be used independently anymore. For example, if the user wants to include only videos files, regardless of file extensions, it won't work and instead include everything including photos. 

In order to fix this, the extension matching logic should probably be moved further down the pipeline to the exiftool metadata steps, where we get the actual file type (video or image). 

To sum up: solely relying on file extensions (from the file system) for filtering doesn't work. :wink:

**Comment by rprieto** [comment]
Thanks for adding more detail.

Here is a use case to think about: people storing Raw + Jpeg side by side, but setting `--include-raw-photos false`. We have to think thumbsup works with a local database of all tracked files. If we include all extensions we have to choose:

* Don’t add raw files to the database: they will get re-parsed on every run
* Add the raw files to the database: it will double in size

The challenge gets even bigger for people running thumbsup on a folder with non-media files, for example to generate a gallery of all photos in their home folder. This is the challenge with

> In order to fix this, the extension matching logic should probably be moved further down the pipeline to the exiftool metadata steps

That part of the pipeline happens after the database update. Because of this, I want to keep extension-based filtering at least as a default.

I will have a think about making it optional. The good news is that it *would* be compatible with `--include-photos | raw-photos | videos`, because exiftool outputs what the file extension should have been:

```bash
$ exiftool -json -struct file.ext | grep 'FileTypeExtension' 
  "FileTypeExtension": "jpg",
```

This means we can also re-filter when streaming the files from the database, if needed.

**Comment by rprieto** [comment]
Thanks for adding more detail.

Here is a use case to think about: people storing Raw + Jpeg side by side, but setting `--include-raw-photos false`. We have to think thumbsup works with a local database of all tracked files. If we include all extensions we have to choose:

* Don’t add raw files to the database: they will get re-parsed on every run
* Add the raw files to the database: it will double in size

The challenge gets even bigger for people running thumbsup on a folder with non-media files, for example to generate a gallery of all photos in their home folder. This is the challenge with

> In order to fix this, the extension matching logic should probably be moved further down the pipeline to the exiftool metadata steps

That part of the pipeline happens after the database update.

Because Thumbsup is not a generic file explorer, I really want to avoid tracking a database of all files, especially as it contains quite a lot of metadata. It would have to something like `--database-tracking [mediaExtensions | all]` but which starts sounding like a file explorer.

> since solely relying on file extensions (from the file system) for filtering should not be considered reliable

In theory I agree with you. In practice though a lot of software expect files to have the right extensions. The question is how common is it to have a JPG file called SWF, and whether the complexity is worth it
- To maintain in the code
- For users to understand the impact on performance and the different flags

**Comment by rprieto** [comment]
Hi @glubsy. Thanks again for the suggestion. It's unlikely thumbsup will implement this in the near future, so I'm closing the issue for now. I will re-open it if there are more use cases for indexing all files regardless of their extension.
